File: src/bmsx_cpp/machine/devices/vdp/sbx.cpp
#include "machine/devices/vdp/sbx.h"

#include "machine/devices/vdp/vram.h"

namespace bmsx {

void VdpSbxUnit::reset() {
	m_liveFaceWords.fill(0u);
	m_faceWindowWords.fill(0u);
	m_packetFaceWords.fill(0u);
	m_sealFaceWords.fill(0u);
	m_visibleFaceWords.fill(0u);
	m_liveControl = 0u;
	m_faceWindowControl = 0u;
	m_packetControl = 0u;
	m_visibleControl = 0u;
	m_frameDecision.state = VdpSbxFrameState::Idle;
	m_frameDecision.control = 0u;
	m_frameDecision.faultCode = VDP_FAULT_NONE;
	m_frameDecision.faultDetail = 0u;
}

void VdpSbxUnit::writeFaceWindowControl(u32 control) {
	m_faceWindowControl = control;
}

void VdpSbxUnit::writeFaceWindowWord(size_t index, u32 word) {
	m_faceWindowWords[index] = word;
}

void VdpSbxUnit::commitFaceWindow() {
	m_liveControl = m_faceWindowControl;
	m_liveFaceWords = m_faceWindowWords;
}

VdpSbxUnit::FaceWords& VdpSbxUnit::beginPacket(u32 control) {
	m_packetControl = control;
	return m_packetFaceWords;
}

void VdpSbxUnit::commitPacket() {
	m_liveControl = m_packetControl;
	m_liveFaceWords = m_packetFaceWords;
}

VdpSbxFrameDecision VdpSbxUnit::beginFrameSeal() {
	VdpSbxFrameDecision& decision = m_frameDecision;
	decision.state = VdpSbxFrameState::PacketOpen;
	m_sealFaceWords = m_liveFaceWords;
	decision.control = m_liveControl;
	decision.faultCode = VDP_FAULT_NONE;
	decision.faultDetail = 0u;
	return decision;
}

VdpSbxFrameDecision VdpSbxUnit::completeFrameSeal(const VdpSbxFrameResolution& resolution) {
	VdpSbxFrameDecision& decision = m_frameDecision;
	if (resolution.faultCode != VDP_FAULT_NONE) {
		decision.state = VdpSbxFrameState::FrameRejected;
		decision.faultCode = resolution.faultCode;
		decision.faultDetail = resolution.faultDetail;
		return decision;
	}
	decision.state = VdpSbxFrameState::FrameSealed;
	decision.faultCode = VDP_FAULT_NONE;
	decision.faultDetail = 0u;
	return decision;
}

bool VdpSbxUnit::resolveFrameSamplesInto(const VdpVramUnit& vram, u32 control, const FaceWords& faceWords, VdpSkyboxSamples& samples, VdpSbxFrameResolution& resolution) const {
	resolution.faultCode = VDP_FAULT_NONE;
	resolution.faultDetail = 0u;
	if ((control & VDP_SBX_CONTROL_ENABLE) == 0u) {
		return true;
	}
	for (size_t index = 0; index < SKYBOX_FACE_COUNT; ++index) {
		if (!resolveSampleInto(
			vram,
			readSkyboxFaceSourceWord(faceWords, index, SKYBOX_FACE_SLOT_WORD),
			readSkyboxFaceSourceWord(faceWords, index, SKYBOX_FACE_U_WORD),
			readSkyboxFaceSourceWord(faceWords, index, SKYBOX_FACE_V_WORD),
			readSkyboxFaceSourceWord(faceWords, index, SKYBOX_FACE_W_WORD),
			readSkyboxFaceSourceWord(faceWords, index, SKYBOX_FACE_H_WORD),
			samples[index],
			resolution)) {
			return false;
		}
	}
	return true;
}

bool VdpSbxUnit::resolveSampleInto(const VdpVramUnit& vram, u32 slot, u32 u, u32 v, u32 w, u32 h, VdpResolvedBlitterSample& target, VdpSbxFrameResolution& resolution) const {
	resolution.faultCode = VDP_FAULT_NONE;
	resolution.faultDetail = 0u;
	if (slot == VDP_SLOT_SYSTEM) {
		target.source.surfaceId = VDP_RD_SURFACE_SYSTEM;
	} else if (slot == VDP_SLOT_PRIMARY) {
		target.source.surfaceId = VDP_RD_SURFACE_PRIMARY;
	} else if (slot == VDP_SLOT_SECONDARY) {
		target.source.surfaceId = VDP_RD_SURFACE_SECONDARY;
	} else {
		resolution.faultCode = VDP_FAULT_SBX_SOURCE_OOB;
		resolution.faultDetail = slot;
		return false;
	}
	target.source.srcX = u;
	target.source.srcY = v;
	target.source.width = w;
	target.source.height = h;
	if (w == 0u || h == 0u) {
		resolution.faultCode = VDP_FAULT_SBX_SOURCE_OOB;
		resolution.faultDetail = w | (h << 16u);
		return false;
	}
	const VdpSurfaceUploadSlot* surface = vram.findSurface(target.source.surfaceId);
	if (surface == nullptr) {
		resolution.faultCode = VDP_FAULT_SBX_SOURCE_OOB;
		resolution.faultDetail = target.source.surfaceId;
		return false;
	}
	const uint64_t sourceRight = static_cast<uint64_t>(u) + static_cast<uint64_t>(w);
	const uint64_t sourceBottom = static_cast<uint64_t>(v) + static_cast<uint64_t>(h);
	if (sourceRight > surface->surfaceWidth || sourceBottom > surface->surfaceHeight) {
		resolution.faultCode = VDP_FAULT_SBX_SOURCE_OOB;
		resolution.faultDetail = u | (v << 16u);
		return false;
	}
	target.surfaceWidth = surface->surfaceWidth;
	target.surfaceHeight = surface->surfaceHeight;
	target.slot = slot;
	return true;
}

void VdpSbxUnit::presentFrame(u32 control, const FaceWords& faceWords) {
	m_visibleControl = control;
	m_visibleFaceWords = faceWords;
}

void VdpSbxUnit::presentLiveState() {
	m_visibleControl = m_liveControl;
	m_visibleFaceWords = m_liveFaceWords;
}

void VdpSbxUnit::restoreLiveState(u32 control, const FaceWords& faceWords) {
	m_liveControl = control;
	m_liveFaceWords = faceWords;
	m_faceWindowControl = m_liveControl;
	m_faceWindowWords = m_liveFaceWords;
}

u32 readSkyboxFaceSourceWord(const VdpSbxUnit::FaceWords& words, size_t faceIndex, size_t field) {
	return words[faceIndex * SKYBOX_FACE_WORD_STRIDE + field];
}

} // namespace bmsx

File: src/bmsx/machine/devices/vdp/sbx.ts
import {
	SKYBOX_FACE_COUNT,
	SKYBOX_FACE_H_WORD,
	SKYBOX_FACE_SLOT_WORD,
	SKYBOX_FACE_U_WORD,
	SKYBOX_FACE_V_WORD,
	SKYBOX_FACE_W_WORD,
	SKYBOX_FACE_WORD_COUNT,
	SKYBOX_FACE_WORD_STRIDE,
	VDP_FAULT_NONE,
	VDP_FAULT_SBX_SOURCE_OOB,
	VDP_RD_SURFACE_PRIMARY,
	VDP_RD_SURFACE_SECONDARY,
	VDP_RD_SURFACE_SYSTEM,
	VDP_SBX_CONTROL_ENABLE,
	VDP_SLOT_PRIMARY,
	VDP_SLOT_SECONDARY,
	VDP_SLOT_SYSTEM,
} from './contracts';
import type { VdpResolvedBlitterSample } from './blitter';
import type { VdpVramUnit } from './vram';

export const VDP_SBX_PACKET_KIND = 0x12000000;
export const VDP_SBX_PACKET_PAYLOAD_WORDS = 1 + SKYBOX_FACE_WORD_COUNT;
export const VDP_SBX_STATE_IDLE = 0;
export const VDP_SBX_STATE_PACKET_OPEN = 1;
export const VDP_SBX_STATE_FRAME_SEALED = 2;
export const VDP_SBX_STATE_FRAME_REJECTED = 3;

export type VdpSbxFrameState =
	| typeof VDP_SBX_STATE_IDLE
	| typeof VDP_SBX_STATE_PACKET_OPEN
	| typeof VDP_SBX_STATE_FRAME_SEALED
	| typeof VDP_SBX_STATE_FRAME_REJECTED;

export type VdpSbxFrameDecision = {
	state: VdpSbxFrameState;
	control: number;
	faultCode: number;
	faultDetail: number;
};

export type VdpSbxFrameResolution = {
	faultCode: number;
	faultDetail: number;
};

export class VdpSbxUnit {
	private readonly liveFaceWords = new Uint32Array(SKYBOX_FACE_WORD_COUNT);
	private liveControl = 0;
	private readonly faceWindowWords = new Uint32Array(SKYBOX_FACE_WORD_COUNT);
	private faceWindowControl = 0;
	private readonly packetFaceWords = new Uint32Array(SKYBOX_FACE_WORD_COUNT);
	private packetControl = 0;
	private readonly sealFaceWords = new Uint32Array(SKYBOX_FACE_WORD_COUNT);
	private readonly visibleFaceWords = new Uint32Array(SKYBOX_FACE_WORD_COUNT);
	private visibleControl = 0;
	private readonly frameDecision: VdpSbxFrameDecision = {
		state: VDP_SBX_STATE_IDLE,
		control: 0,
		faultCode: VDP_FAULT_NONE,
		faultDetail: 0,
	};

	public reset(): void {
		this.liveFaceWords.fill(0);
		this.faceWindowWords.fill(0);
		this.packetFaceWords.fill(0);
		this.sealFaceWords.fill(0);
		this.visibleFaceWords.fill(0);
		this.liveControl = 0;
		this.faceWindowControl = 0;
		this.packetControl = 0;
		this.visibleControl = 0;
		this.frameDecision.state = VDP_SBX_STATE_IDLE;
		this.frameDecision.control = 0;
		this.frameDecision.faultCode = VDP_FAULT_NONE;
		this.frameDecision.faultDetail = 0;
	}

	public writeFaceWindowControl(control: number): void {
		this.faceWindowControl = control >>> 0;
	}

	public writeFaceWindowWord(index: number, word: number): void {
		this.faceWindowWords[index] = word >>> 0;
	}

	public commitFaceWindow(): void {
		this.liveControl = this.faceWindowControl >>> 0;
		this.liveFaceWords.set(this.faceWindowWords);
	}

	public beginPacket(control: number): Uint32Array {
		this.packetControl = control >>> 0;
		return this.packetFaceWords;
	}

	public commitPacket(): void {
		this.liveControl = this.packetControl >>> 0;
		this.liveFaceWords.set(this.packetFaceWords);
	}

	public beginFrameSeal(): VdpSbxFrameDecision {
		const decision = this.frameDecision;
		decision.state = VDP_SBX_STATE_PACKET_OPEN;
		this.sealFaceWords.set(this.liveFaceWords);
		decision.control = this.liveControl >>> 0;
		decision.faultCode = VDP_FAULT_NONE;
		decision.faultDetail = 0;
		return decision;
	}

	public completeFrameSeal(resolution: VdpSbxFrameResolution): VdpSbxFrameDecision {
		const decision = this.frameDecision;
		if (resolution.faultCode !== VDP_FAULT_NONE) {
			decision.state = VDP_SBX_STATE_FRAME_REJECTED;
			decision.faultCode = resolution.faultCode;
			decision.faultDetail = resolution.faultDetail;
			return decision;
		}
		decision.state = VDP_SBX_STATE_FRAME_SEALED;
		decision.faultCode = VDP_FAULT_NONE;
		decision.faultDetail = 0;
		return decision;
	}

	public presentFrame(control: number, faceWords: Uint32Array): void {
		this.visibleControl = control >>> 0;
		this.visibleFaceWords.set(faceWords);
	}

	public presentLiveState(): void {
		this.visibleControl = this.liveControl >>> 0;
		this.visibleFaceWords.set(this.liveFaceWords);
	}

	public captureLiveFaceWords(): number[] {
		const words = new Array<number>(SKYBOX_FACE_WORD_COUNT);
		for (let index = 0; index < SKYBOX_FACE_WORD_COUNT; index += 1) {
			words[index] = this.liveFaceWords[index] >>> 0;
		}
		return words;
	}

	public restoreLiveState(control: number, faceWords: ArrayLike<number>): void {
		this.liveControl = control >>> 0;
		this.liveFaceWords.set(faceWords);
		this.faceWindowControl = this.liveControl >>> 0;
		this.faceWindowWords.set(this.liveFaceWords);
	}

	public resolveFrameSamplesInto(vram: VdpVramUnit, control: number, faceWords: Uint32Array, samples: VdpResolvedBlitterSample[], resolution: VdpSbxFrameResolution): boolean {
		resolution.faultCode = VDP_FAULT_NONE;
		resolution.faultDetail = 0;
		if ((control & VDP_SBX_CONTROL_ENABLE) === 0) {
			return true;
		}
		for (let index = 0; index < SKYBOX_FACE_COUNT; index += 1) {
			if (!this.resolveSampleInto(
				vram,
				readSkyboxFaceSource(faceWords, index, SKYBOX_FACE_SLOT_WORD),
				readSkyboxFaceSource(faceWords, index, SKYBOX_FACE_U_WORD),
				readSkyboxFaceSource(faceWords, index, SKYBOX_FACE_V_WORD),
				readSkyboxFaceSource(faceWords, index, SKYBOX_FACE_W_WORD),
				readSkyboxFaceSource(faceWords, index, SKYBOX_FACE_H_WORD),
				samples[index]!,
				resolution,
			)) {
				return false;
			}
		}
		return true;
	}

	private resolveSampleInto(
		vram: VdpVramUnit,
		slot: number,
		u: number,
		v: number,
		w: number,
		h: number,
		target: VdpResolvedBlitterSample,
		resolution: VdpSbxFrameResolution,
	): boolean {
		resolution.faultCode = VDP_FAULT_NONE;
		resolution.faultDetail = 0;
		const source = target.source;
		if (slot === VDP_SLOT_SYSTEM) {
			source.surfaceId = VDP_RD_SURFACE_SYSTEM;
		} else if (slot === VDP_SLOT_PRIMARY) {
			source.surfaceId = VDP_RD_SURFACE_PRIMARY;
		} else if (slot === VDP_SLOT_SECONDARY) {
			source.surfaceId = VDP_RD_SURFACE_SECONDARY;
		} else {
			resolution.faultCode = VDP_FAULT_SBX_SOURCE_OOB;
			resolution.faultDetail = slot;
			return false;
		}
		source.srcX = u;
		source.srcY = v;
		source.width = w;
		source.height = h;
		if (w === 0 || h === 0) {
			resolution.faultCode = VDP_FAULT_SBX_SOURCE_OOB;
			resolution.faultDetail = (w | (h << 16)) >>> 0;
			return false;
		}
		const surface = vram.findSurface(source.surfaceId);
		if (surface === null) {
			resolution.faultCode = VDP_FAULT_SBX_SOURCE_OOB;
			resolution.faultDetail = source.surfaceId;
			return false;
		}
		if (u + w > surface.surfaceWidth || v + h > surface.surfaceHeight) {
			resolution.faultCode = VDP_FAULT_SBX_SOURCE_OOB;
			resolution.faultDetail = (u | (v << 16)) >>> 0;
			return false;
		}
		target.surfaceWidth = surface.surfaceWidth;
		target.surfaceHeight = surface.surfaceHeight;
		target.slot = slot;
		return true;
	}

	public get visibleEnabled(): boolean {
		return (this.visibleControl & VDP_SBX_CONTROL_ENABLE) !== 0;
	}

	public get liveControlWord(): number {
		return this.liveControl >>> 0;
	}

	public get visibleFaceState(): Uint32Array {
		return this.visibleFaceWords;
	}

	public get liveFaceState(): Uint32Array {
		return this.liveFaceWords;
	}

	public get sealFaceState(): Uint32Array {
		return this.sealFaceWords;
	}
}

export function readSkyboxFaceSource(words: Uint32Array, faceIndex: number, field: number): number {
	return words[faceIndex * SKYBOX_FACE_WORD_STRIDE + field] >>> 0;
}
